cl/alerts/tasks.py

```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
+        except UserProfile.DoesNotExist:
+            recap_email_user_does_not_exist_list.append(email_address)
+            continue
+        email_address = user_profile.user.email
+        docket_alert_exist = DocketAlert.objects.filter(
+            docket_id=d_pk, user=user_profile.user
+        ).exists()
+        if docket_alert_exist:
+            # If a docket alert exists for this @recap.email user, avoid
+            # sending the first email
+            continue
+        if user_profile.auto_subscribe:
+            # First time recipient, auto_subscribe True
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk, user=user_profile.user
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_auto_subscribe_list.append(recipient_tuple)
+        else:
+            # First time recipient, auto_subscribe False
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk,
+                user=user_profile.user,
+                alert_type=DocketAlert.UNSUBSCRIPTION,
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_no_auto_subscribe_list.append(
+                recipient_tuple
+            )
+    return (
+        docket_alert_recipients_list,
+        recap_email_first_time_auto_subscribe_list,
+        recap_email_first_time_no_auto_subscribe_list,
+        recap_email_user_does_not_exist_list,
+    )
+
+
 def make_alert_messages(
     d: Docket,
     new_des: QuerySet,
-    email_addresses: "ValuesQuerySet[User, Optional[str]]",  # type: ignore
-    recap_email_recipients: list[str] = [],
-) -> List[EmailMultiAlternatives]:
+    recipients: list[tuple[str, str]],
+    first_email: bool,
+    auto_subscribe: bool,
+) -> list[EmailMultiAlternatives]:
     """Make docket alert messages that can be sent to users
 
     :param d: The docket to work on
     :param new_des: The new docket entries
-    :param email_addresses: A list of user email addresses to send to
-    :param recap_email_recipients: The recap_email addresses if needed to send
-    the first case-user notification email.
+    :param recipients: A list of user email and secret_key tuples to send to
+    :param first_email: Whether this is the first email for this docket
+    :param auto_subscribe: Whether this is the first time the user has received
+    a notification for this docket
     :return: A list of email messages to send
     """
 
     case_name = trunc(best_case_name(d), 100, ellipsis="...")
-    messages = []
     txt_template = loader.get_template("docket_alert_email.txt")
     html_template = loader.get_template("docket_alert_email.html")
     subject_template = loader.get_template("docket_alert_subject.txt")
     subject_context = {
         "docket": d,
         "count": new_des.count(),
```

Comment:
You'll save a query by making this `len(new_des)`. Just saying!

Reply:
Changed!